File: apps/macos/Sources/Displays/DisplayDescriptor.swift
import AppKit

/// A snapshot of one connected display, keyed by an identifier that survives
/// reordering, sleep, and reconnection.
nonisolated struct DisplayDescriptor: Identifiable, Hashable, Sendable {
    let id: CGDirectDisplayID
    let name: String
    let frame: CGRect
    let backingScaleFactor: CGFloat
    let isPrimary: Bool

    /// Pixel resolution, useful when explaining energy impact to the user.
    var pixelSize: CGSize {
        CGSize(width: frame.width * backingScaleFactor, height: frame.height * backingScaleFactor)
    }

    var resolutionDescription: String {
        "\(Int(pixelSize.width)) × \(Int(pixelSize.height))"
    }
}

extension NSScreen {
    /// The stable Core Graphics display ID. Never persist `NSScreen.screens`
    /// indexes: display ordering changes when monitors are docked.
    var displayID: CGDirectDisplayID? {
        let key = NSDeviceDescriptionKey("NSScreenNumber")
        return (deviceDescription[key] as? NSNumber)?.uint32Value
    }

    var displayDescriptor: DisplayDescriptor? {
        guard let displayID else { return nil }
        return DisplayDescriptor(
            id: displayID,
            name: localizedName,
            frame: frame,
            backingScaleFactor: backingScaleFactor,
            isPrimary: CGDisplayIsMain(displayID) != 0
        )
    }
}

File: apps/windows/src/Wallpache.App/Displays/DisplayDescriptor.cs
using System;

namespace Wallpache.App.Displays;

/// <summary>
/// A snapshot of one connected display, keyed by an identifier that survives
/// reordering, sleep, and reconnection.
///
/// <see cref="Id"/> is the monitor's device interface path where Windows can
/// supply one, so docking a laptop does not shuffle wallpaper assignments the
/// way an index or a bare <c>\\.\DISPLAY1</c> name would.
/// </summary>
public sealed record DisplayDescriptor(
    string Id,
    IntPtr MonitorHandle,
    string DeviceName,
    string Name,
    int X,
    int Y,
    int Width,
    int Height,
    bool IsPrimary,
    double ScaleFactor)
{
    /// <summary>Pixel resolution, useful when explaining energy impact to the user.</summary>
    public string ResolutionDescription => $"{Width} × {Height}";

    /// <summary>
    /// Compares everything that would require a session to be moved or resized.
    /// The monitor handle is deliberately excluded: it changes on every
    /// reconfiguration without meaning the display did.
    /// </summary>
    public bool IsEquivalentTo(DisplayDescriptor other) =>
        Id == other.Id
        && DeviceName == other.DeviceName
        && Name == other.Name
        && X == other.X
        && Y == other.Y
        && Width == other.Width
        && Height == other.Height
        && IsPrimary == other.IsPrimary;
}
